modules/apps/router/keeper/keeper.go

```diff
@@ -0,0 +1,223 @@
+package keeper
+
+import (
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/tendermint/tendermint/libs/log"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	paramtypes "github.com/cosmos/cosmos-sdk/x/params/types"
+	"github.com/cosmos/ibc-go/modules/apps/router/types"
+	transfertypes "github.com/cosmos/ibc-go/modules/apps/transfer/types"
+	clienttypes "github.com/cosmos/ibc-go/modules/core/02-client/types"
+	channeltypes "github.com/cosmos/ibc-go/modules/core/04-channel/types"
+	host "github.com/cosmos/ibc-go/modules/core/24-host"
+	coretypes "github.com/cosmos/ibc-go/modules/core/types"
+)
+
+// Keeper defines the IBC fungible transfer keeper
+type Keeper struct {
+	storeKey   sdk.StoreKey
+	cdc        codec.BinaryCodec
+	paramSpace paramtypes.Subspace
+
+	transferKeeper types.TransferKeeper
+	bankKeeper     types.BankKeeper
+}
+
+// NewKeeper creates a new 29-fee Keeper instance
+func NewKeeper(
+	cdc codec.BinaryCodec, key sdk.StoreKey, paramSpace paramtypes.Subspace,
+	transferKeeper types.TransferKeeper, bankKeeper types.BankKeeper,
+) Keeper {
+
+	return Keeper{
+		cdc:            cdc,
+		storeKey:       key,
+		transferKeeper: transferKeeper,
+		paramSpace:     paramSpace,
+		bankKeeper:     bankKeeper,
+	}
+}
+
+// Logger returns a module-specific logger.
+func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", "x/"+host.ModuleName+"-"+types.ModuleName)
+}
+
+func (k Keeper) OnRecvPacket(ctx sdk.Context, packet channeltypes.Packet, data transfertypes.FungibleTokenPacketData, receiver sdk.AccAddress, finalDest, port, channel string) error {
```

Comment:
Do not redo the transfer logic here. Instead just use the underlying app, recalculate the denom, and call `ForwardPacket`

Reply:
Ok! I think I know what to do now.